### PR TITLE
Message filtering

### PR DESCRIPTION
**Pull request by MatejFranceskin**
Add Filter field to endpoint configuration to limit the messages that are sent to particular endpoint and improve performance.
For example XYZ is interested only in messages 24,33,76,320,321,323.
This configuration filters all other messages so that they're not sent to XYZ and don't need to be parsed.

[UdpEndpoint XYZ]
Mode = Normal
Address = 127.0.0.1
Port = 14530
Filter = 24,33,76,320,321,323

### CODE REVIEW

File: src/mavlink-router/endpoint.cpp [hunk 1]
```diff
@@ -256,6 +257,7 @@ int Endpoint::read_msg(struct buffer *pbuf, int *target_sysid, int *target_compi
                 *target_compid = 0;
             }
         }
+        *msg_id = msg_entry->msgid;
```
Comment (bkueng) [inline src/mavlink-router/endpoint.cpp]: Do you see any case where it is not already equal?

**Review by bkueng — COMMENTED** [review]
Looks good from a first look at it.
Some points that come to my mind (not necessarily to implement here):
- we could allow to specify MAVLink message names instead of ID's, which would be nicer
- have you thought about RX filters? That would be helpful for very simple firewall (e.g. don't allow commands for certain endpoints)
- you implemented a whitelist, but I see a blacklist as helpful as well for certain cases as well (like high-rate IMU data sending only to a specific endpoint, and blacklist from all others). Alternatively (or in addition) specifying the maximum rate for a message might be useful as well.

**Comment by Jaeyoung-Lim** [comment]
Really really cool!

File: src/mavlink-router/endpoint.cpp (hunk 1 above)
Comment (MatejFranceskin) [reply src/mavlink-router/endpoint.cpp]: Right - that is not needed!

**Comment by MatejFranceskin** [comment]
I added also option to blacklist messages (exclusive filter).

File: src/mavlink-router/main.cpp [hunk 2]
```diff
@@ -660,11 +677,15 @@ static int parse_confs(ConfFile &conf)
...
+        char *filter_exc;
     };
     static const ConfFile::OptionsTable option_table_udp[] = {
-        {"address", true,   ConfFile::parse_str_dup,    OPTIONS_TABLE_STRUCT_FIELD(option_udp, addr)},
-        {"mode",    true,   parse_mode,                 OPTIONS_TABLE_STRUCT_FIELD(option_udp, eavesdropping)},
-        {"port",    false,  ConfFile::parse_ul,         OPTIONS_TABLE_STRUCT_FIELD(option_udp, port)},
+        {"address",    true,   ConfFile::parse_str_dup,    OPTIONS_TABLE_STRUCT_FIELD(option_udp, addr)},
+        {"mode",       true,   parse_mode,                 OPTIONS_TABLE_STRUCT_FIELD(option_udp, eavesdropping)},
+        {"port",       false,  ConfFile::parse_ul,         OPTIONS_TABLE_STRUCT_FIELD(option_udp, port)},
+        {"filter",     false,  ConfFile::parse_str_dup,    OPTIONS_TABLE_STRUCT_FIELD(option_udp, filter_inc)},
```
Comment (bkueng) [inline src/mavlink-router/main.cpp]: I suggest to rename to something like (also `filter_exc` -> `tx_filter_exc`), so we can add rx later on:
```suggestion
        {"tx_filter_inc",     false,  ConfFile::parse_str_dup,    OPTIONS_TABLE_STRUCT_FIELD(option_udp, filter_inc)},
```

File: src/mavlink-router/main.cpp [hunk 3]
```diff
@@ -660,11 +677,15 @@ static int parse_confs(ConfFile &conf)
         char *addr;
         bool eavesdropping;
         unsigned long port;
+        char *filter_inc;
```
Comment (bkueng) [inline src/mavlink-router/main.cpp]: Can you also add them for TCP and UART for consistency?

**Review by bkueng — COMMENTED** [review]
Thanks for the update.
Can you document the filters in the example config file?

**Comment by Jaeyoung-Lim** [comment]
@MatejFranceskin What is the status of this? Can we merge this and test it on the distro side?

File: src/mavlink-router/main.cpp (hunk 3 above)
Comment (MatejFranceskin) [reply src/mavlink-router/main.cpp]: Added.

**Comment by MatejFranceskin** [comment]
@Jaeyoung-Lim @bkueng 
I completely forgot about this. Now I rebased it and implemented review suggestions.

File: src/mavlink-router/main.cpp [hunk 4]
```diff
@@ -223,23 +241,39 @@ static int add_endpoint_address(const char *name, size_t name_len, const char *i
     conf->type = Udp;
     conf->port = ULONG_MAX;
 
-    if (name) {
+    if (!conf->name && name) {
```
Comment (bkueng) [inline src/mavlink-router/main.cpp:244]: This change looks unrelated, and not necessary as `config` is zero-initialized.

File: src/mavlink-router/main.cpp [hunk 5]
```diff
@@ -223,23 +241,39 @@ static int add_endpoint_address(const char *name, size_t name_len, const char *i
...
         conf->name = strndup(name, name_len);
         if (!conf->name) {
             ret = -ENOMEM;
             goto fail;
         }
     }
 
-    conf->address = strdup(ip);
+    if (ip) {
+        free(conf->address);
```
Comment (bkueng) [inline src/mavlink-router/main.cpp:253]: `conf->address` is NULL here.

File: src/mavlink-router/mainloop.cpp [hunk 6]
```diff
@@ -497,6 +497,21 @@ bool Mainloop::add_endpoints(Mainloop &mainloop, struct options *opt)
                     return false;
             }
 
+            if (conf->tx_filter_inc) {
```
Comment (bkueng) [inline src/mavlink-router/mainloop.cpp:500]: Can you move the parsing into a separate method and reuse it below?

File: src/mavlink-router/main.cpp [hunk 7]
```diff
@@ -343,6 +378,22 @@ static int add_uart_endpoint(const char *name, size_t name_len, const char *uart
...
+    if (tx_filter_inc) {
+        conf->tx_filter_inc = strdup(tx_filter_inc);
+        if (!conf->tx_filter_inc) {
+            ret = -ENOMEM;
+            goto fail;
+        }
+    }
+
+    if (tx_filter_exc) {
+        conf->tx_filter_exc = strdup(tx_filter_exc);
```
Comment (bkueng) [inline src/mavlink-router/main.cpp:390]: Where is this freed?

**Review by bkueng — COMMENTED** [review]
Thanks Matej.

Can you document the filters in the example config file?

**Comment by jkflying** [comment]
This is now merged via upstream, correct? Can we close this PR?

**Comment by MatejFranceskin** [comment]
True. Closing.
